hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/ProtobufRpcEngine2.java

```diff
@@ -629,6 +630,7 @@ private RpcWritable call(RPC.Server server,
   // which uses the rpc header.  in the normal case we want to defer decoding
   // the rpc header until needed by the rpc engine.
   static class RpcProtobufRequest extends RpcWritable.Buffer {
+    private final ReentrantLock lock = new ReentrantLock();
```

Comment:
I guess re-entrant lock is unnecessary.
adding `synchronized` to `getRequestHeader()` method and calling this method instead of all direct usages of `requestHeader` would be sufficient. 
This would ensure that `requestHeader` will never be null. So unnecessary null check can be removed.

Remember, this is a critical path, any extra performance impacts, however small it is would impact the overall RPC performance.